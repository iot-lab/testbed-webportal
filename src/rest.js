import axios from 'axios'

const API_URL = `https://${process.env.VUE_APP_IOTLAB_HOST}/api/`

export const iotlab = {

  api: axios.create({baseURL: API_URL, auth: JSON.parse(localStorage.getItem('apiAuth') || '{}')}),

  create (username, password) {
    this.api = axios.create({
      baseURL: API_URL,
      auth: {
        username: username,
        password: password,
      },
    })
  },

  // USERS API

  async signup (user, nbUsers) {
    if (nbUsers) {
      // create multiple user accounts
      await iotlab.api.post(`/users?nbusers=${nbUsers}`, user)
    } else {
      // create single user account
      await iotlab.api.post('/users', user)
    }
  },

  async activateAccount (hash) {
    await iotlab.api.post('/users/activate/', {'hash': hash})
  },

  async deleteUser (login = undefined) {
    if (login) {
      await iotlab.api.delete(`/users/${login}`)
    } else {
      await iotlab.api.delete(`/user`)
    }
  },

  async resetPassword (email) {
    await iotlab.api.post(`/users/reset_password`, {'email': email})
  },

  async changePassword (oldPwd, newPwd, confirmPwd) {
    await iotlab.api.put('/user/password', {
      'old_password': oldPwd,
      'new_password': newPwd,
      'confirm_new_password': confirmPwd,
    })
  },

  async getSSHkeys () {
    return iotlab.api.get('/user/keys').then(resp => resp.data.sshkeys)
  },

  async modifySSHkeys (keys) {
    return iotlab.api.post('/user/keys', {'sshkeys': keys})
  },

  async deleteSSHkey (id) {
    return iotlab.api.delete(`/user/keys/${id}`)
  },

  async getUsers ({status, group, search} = {}) {
    let params = {}
    if (status) {
      params.status = status
    }
    if (group) {
      params.group = group
    }
    if (search) {
      params.search = search
    }
    return iotlab.api.get('/users', {params: params}).then(resp => resp.data)
  },

  async getUserInfo (login = undefined) {
    if (login) {
      return iotlab.api.get(`/users/${login}`).then(resp => resp.data)
    } else {
      return iotlab.api.get('/user').then(resp => resp.data)
    }
  },

  async setUserInfo (user, login = undefined) {
    if (login) {
      return iotlab.api.put(`/users/${login}`, user)
    } else {
      return iotlab.api.put('/user', user)
    }
  },

  async activateUser (login) {
    return iotlab.api.put(`/users/${login}`, {'status': 'active'})
  },

  async deactivateUser (login) {
    return iotlab.api.put(`/users/${login}`, {'status': 'inactive'})
  },

  async setUserGroups (login, groups) {
    return iotlab.api.put(`/users/${login}`, {'groups': groups})
  },

  async sendMail (recipients, subject, message) {
    await iotlab.api.post('/users/email', {
      recipients: recipients,
      subject: subject,
      body: message,
    })
  },

  // EXPERIMENT API

  async getUserExperimentsCount () {
    return iotlab.api.get('/experiments/total').then(resp => resp.data)
  },

  async getAllExperimentsCount (user) {
    if (user) {
      return iotlab.api.get(`/experiments/total/users/${user}`).then(resp => resp.data)
    } else {
      return iotlab.api.get(`/experiments/total/all`).then(resp => resp.data)
    }
  },

  async getAllExperiments ({limit, offset, user, state = 'Terminated,Error,Running,Finishing,Resuming,toError,Waiting,Launching,Hold,toLaunch,toAckReservation,Suspended'} = {}) {
    let params = {}
    if (state) {
      params.state = state
    }
    if (limit) {
      params.limit = Math.min(limit, 500) // IoT-LAB api restrict queries to 500 entries
    }
    if (offset) {
      params.offset = offset
    }
    if (user === '@self') {
      return iotlab.api.get('/experiments', {params: params}).then(resp => resp.data.items)
    } else if (user) {
      return iotlab.api.get(`/experiments/users/${user}`, {params: params}).then(resp => resp.data.items)
    } else {
      return iotlab.api.get('/experiments/all', {params: params}).then(resp => resp.data.items)
    }
  },

  async getRunningExperiments () {
    return iotlab.api.get('/experiments/running').then(resp => resp.data.items)
  },

  async getExperiment (id) {
    return iotlab.api.get(`/experiments/${id}`).then(resp => resp.data)
  },

  async getExperimentArchive (id) {
    return iotlab.api.get(`/experiments/${id}/data`, {responseType: 'arraybuffer'}).then(resp => resp.data)
  },

  async getExperimentDeployment (id) {
    return iotlab.api.get(`/experiments/${id}/deployment`).then(resp => resp.data)
  },

  async stopExperiment (id) {
    return iotlab.api.delete(`/experiments/${id}`)
  },

  async submitExperiment ({type, name, duration, reservation, nodes, firmwareassociations, firmwares, profileassociations, profiles} = {}) {
    const formData = new FormData()
    formData.append('experiment.json', JSON.stringify({
      type: type,
      duration: duration,
      reservation: reservation,
      name: name,
      nodes: nodes,
      firmwareassociations: firmwareassociations,
      profileassociations: profileassociations,
    }))
    // attached firmware names should be unique
    let attachedFirmwares = []
    for (let asso of firmwareassociations || []) {
      if (asso.firmwarename in firmwares && !attachedFirmwares.includes(asso.firmwarename)) {
        formData.append(asso.firmwarename, new Blob([firmwares[asso.firmwarename]], {type: 'application/octet-stream'}), asso.firmwarename)
        attachedFirmwares.push(asso.firmwarename)
      }
    }
    return iotlab.api.post('/experiments', formData).then(resp => resp.data)
  },

  async checkFirmware (fileString) {
    const formData = new FormData()
    formData.append('firmware.bin', fileString)
    return iotlab.api.post('/firmwares/checker', formData).then(resp => resp.data)
  },

  async flashFirmware (id, nodes, fileString) {
    const formData = new FormData()
    formData.append(`experiment${id}.json`, JSON.stringify(nodes))
    formData.append('firmware.bin', new Blob([fileString], {type: 'application/octet-stream'}), 'firmware.bin')
    return iotlab.api.post(`/experiments/${id}/nodes/flash`, formData).then(resp => resp.data)
  },

  async flashResourcesFirmware (id, nodes, name) {
    return iotlab.api.post(`/experiments/${id}/nodes/flash/${name}`, nodes).then(resp => resp.data)
  },

  async updateMonitoring (id, nodes, name) {
    return iotlab.api.post(`/experiments/${id}/nodes/monitoring/${name}`, nodes).then(resp => resp.data)
  },

  async updateExperimentMobilityCircuit (id, nodes, name) {
    return iotlab.api.post(`/experiments/${id}/robots/mobility/${name}`, nodes).then(resp => resp.data)
  },

  async reloadExperiment (id) {
    return iotlab.api.post(`/experiments/${id}/reload`, {}).then(resp => resp.data)
  },

  // USER GROUPS

  async getUserGroups () {
    return iotlab.api.get('/groups').then(resp => resp.data)
  },

  async createGroup (name) {
    return iotlab.api.post(`/groups`, {name: name})
  },

  async deleteGroup (group) {
    return iotlab.api.delete(`/groups/${group}`)
  },

  // OTHER API

  async getSites () {
    return iotlab.api.get('/sites').then(resp => resp.data.items)
  },

  async getSitesDetails () {
    return iotlab.api.get('/sites/details').then(resp => resp.data.items)
  },

  async getNodes () {
    return iotlab.api.get('/nodes').then(resp => resp.data.items)
  },

  async updateNodesProperties () {
    return iotlab.api.put('/nodes')
  },

  async updateNodesPropertiesOldApi () {
    const OLD_API = `https://${process.env.VUE_APP_IOTLAB_HOST}/rest/`
    const oldApi = axios.create({baseURL: OLD_API, auth: JSON.parse(localStorage.getItem('apiAuth') || '{}')})
    return oldApi.post('/admin/nodes')
  },

  async getNodesIds () {
    return iotlab.api.get('/nodes/ids').then(resp => resp.data.items)
  },

  async getExperimentNodes (id) {
    return iotlab.api.get(`/experiments/${id}/nodes`).then(resp => resp.data.items)
  },

  async getExperimentToken (id) {
    return iotlab.api.get(`/experiments/${id}/token`).then(resp => resp.data.token)
  },

  async sendNodesCommand (id, cmd, nodes) {
    // cmd in [start, stop, reset, flash-idle, profile-reset, debug-start, debug-stop]
    return iotlab.api.post(`/experiments/${id}/nodes/${cmd}`, nodes).then(resp => resp.data)
  },

  // STORES (firmwares, profiles, etc)

  async getMobilityCircuits () {
    return iotlab.api.get('/mobilities/circuits').then(resp => resp.data.items)
  },

  async getMobilityCircuit (name) {
    return iotlab.api.get(`/mobilities/circuits/${name}`).then(resp => resp.data)
  },

  async getRobotsSiteMapImage (site) {
    return iotlab.api.get(`/robots/${site}/map/image`, {
      responseType: 'arraybuffer',
      headers: {
        'Access-Control-Allow-Origin': '*',
      },
    }).then(resp => Buffer.from(resp.data, 'binary').toString('base64')
    )
  },

  async getRobotsSiteMapConfig (site) {
    return iotlab.api.get(`/robots/${site}/map/config`).then(resp => resp.data)
  },

  async createMobilityCircuit (circuit) {
    return iotlab.api.post(`/mobilities/circuits`, circuit)
  },

  async updateMobilityCircuit (name, circuit) {
    return iotlab.api.put(`/mobilities/circuits/${name}`, circuit)
  },

  async deleteMobilityCircuit (name) {
    return iotlab.api.delete(`/mobilities/circuits/${name}`)
  },

  async getMonitoringProfiles () {
    return iotlab.api.get('/monitoring').then(resp => resp.data)
  },

  async getMonitoringProfile (name) {
    return iotlab.api.get(`/monitoring/${name}`).then(resp => resp.data)
  },

  async updateMonitoringProfile (name, profile) {
    return iotlab.api.put(`/monitoring/${name}`, profile)
  },

  async createMonitoringProfile (profile) {
    return iotlab.api.post(`/monitoring`, profile)
  },

  async deleteMonitoringProfile (name) {
    return iotlab.api.delete(`/monitoring/${name}`)
  },

  async getFirmwares () {
    return iotlab.api.get('/firmwares').then(resp => resp.data)
  },

  async getFirmware (name) {
    return iotlab.api.get(`/firmwares/${name}`).then(resp => resp.data)
  },

  async getFirmwareFile (name) {
    return iotlab.api.get(`/firmwares/${name}/file`, {responseType: 'arraybuffer'}).then(resp => resp.data)
  },

  async updateFirmware (name, firmware, fileString) {
    const formData = new FormData()
    formData.append(`firmware.json`, JSON.stringify(firmware))
    if (fileString) {
      formData.append('firmware.bin', new Blob([fileString], {type: 'application/octet-stream'}), 'firmware.bin')
    }
    return iotlab.api.put(`/firmwares/${name}`, formData).then(resp => resp.data)
  },

  async createFirmware (firmware, fileString) {
    const formData = new FormData()
    formData.append(`firmware.json`, JSON.stringify(firmware))
    formData.append('firmware.bin', new Blob([fileString], {type: 'application/octet-stream'}), 'firmware.bin')
    return iotlab.api.post(`/firmwares`, formData).then(resp => resp.data)
  },

  async deleteFirmware (name) {
    return iotlab.api.delete(`/firmwares/${name}`)
  },
}
