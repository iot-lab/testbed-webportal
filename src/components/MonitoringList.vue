<template>
  <div>
    <router-link :to="{name: 'newMonitoring'}" class="btn btn-sm btn-success float-right"><i class="fa fa-plus"></i> New profile</router-link>
    <h5>Monitoring profiles</h5>
    <table class="table table-striped table-sm mt-3">
      <thead>
        <tr>
          <th class="cursor" title="sort by name" @click="sortBy(p => p.profilename)">Profile</th>
          <th class="cursor" title="sort by archi" @click="sortBy(p => nodeArchi(p))">Archi</th>
          <th>Monitor</th>
        </tr>
      </thead>
      <tbody>
        <tr v-for="profile in store.profiles" v-if="filterByArchi(profile)">
          <td>
            <a href="#" @click.prevent="select(profile)">{{profile.profilename}}</a>
          </td>
          <td>{{nodeArchi(profile)}}</td>
          <td>
            <template v-if="profile.consumption && (profile.consumption.current || profile.consumption.voltage || profile.consumption.power)">
              <i class="fa fa-bolt" title="consumption"></i>
              <span v-if="profile.consumption.current"> current </span>
              <span v-if="profile.consumption.voltage"> voltage </span>
              <span v-if="profile.consumption.power"> power </span>
              <span class="text-muted" title="period, average">({{profile.consumption.period}}Âµs, {{profile.consumption.average}})</span>
            </template>
            <template v-if="profile.radio">
              <i class="fa fa-wifi" title="radio"></i> {{profile.radio.mode}}
              <span class="text-muted" title="channel">({{profile.radio.channels.join(',')}})</span>
            </template>
          </td>
        </tr>
        <tr v-if="store.profiles.length === 0">
          <td colspan="3" class="font-italic bg-light">
            <router-link :to="{name: 'newMonitoring'}" class="fsont-italic">create a monitoring profile</router-link>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</template>

<script>
import { iotlab } from '@/rest'
import store from '@/store'

export default {
  name: 'MonitoringList',

  props: {
    archi: {
      // Display monitoring profiles filtered by archi
      type: String,
      default: '',
    },
  },

  data () {
    return {
      store: store,
    }
  },

  async created () {
    try {
      this.store.profiles = (await iotlab.getMonitoringProfiles()).sort((a, b) => a.profilename.localeCompare(b.profilename))
    } catch (err) {
      this.$notify({text: 'Failed to load profiles', type: 'error'})
    }
  },

  methods: {
    sortBy (func) {
      // sort by func() then by profile name
      this.store.profiles = this.store.profiles.sort((a, b) => func(a) === func(b) ? a.profilename.localeCompare(b.profilename) : func(a).localeCompare(func(b)))
    },
    nodeArchi (profile) {
      return profile.nodearch.replace('custom', 'other')
    },
    filterByArchi (profile) {
      switch (this.archi) {
        case '': return true
        case 'a8': return profile.nodearch === 'a8'
        case 'm3': return profile.nodearch === 'm3'
        default: return profile.nodearch === 'custom'
      }
    },
    select (profile) {
      this.$emit('select', profile.profilename)
    },
  },
}
</script>

<style scoped>

</style>
