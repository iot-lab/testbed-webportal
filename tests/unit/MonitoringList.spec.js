import { shallowMount, createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import Notifications from 'vue-notification'
import MonitoringList from '@/components/MonitoringList'
import { iotlab } from '@/rest'

jest.mock('@/rest')
iotlab.getMonitoringProfiles = jest.fn().mockResolvedValue([{"profilename":"a8prof","consumption":{"average":4,"period":8244,"current":true,"power":false,"voltage":false},"power":"dc","nodearch":"a8"},{"profilename":"conso","consumption":{"average":4,"period":8244,"current":true,"power":true,"voltage":true},"power":"dc","nodearch":"m3"},{"profilename":"radio","power":"dc","nodearch":"custom","radio":{"mode":"rssi","period":1000,"channels":[11,13],"num_per_channel":16}}])

const router = new VueRouter()
const localVue = createLocalVue()
localVue.use(VueRouter)
localVue.use(Notifications)

describe('MonitoringList.vue', () => {

  it('should match snapshot', async () => {
    const wrapper = await shallowMount(MonitoringList, {localVue, router})
    await wrapper.vm.$nextTick() // wait for promise to resolve
    expect(wrapper.element).toBeDefined()
    expect(wrapper.element).toMatchSnapshot()
  })

})
