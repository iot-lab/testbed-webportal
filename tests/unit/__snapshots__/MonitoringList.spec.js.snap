// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonitoringList.vue should match snapshot 1`] = `
<div>
  <vuecomponent-stub
    class="btn btn-sm btn-success float-right"
    to="[object Object]"
  />
   
  <h5>
    Monitoring profiles
  </h5>
   
  <table
    class="table table-striped table-sm mt-3"
  >
    <thead>
      <tr>
        <th
          class="cursor"
          title="sort by name"
        >
          Profile
        </th>
         
        <th
          class="cursor"
          title="sort by archi"
        >
          Archi
        </th>
         
        <th>
          Monitor
        </th>
      </tr>
    </thead>
     
    <tbody>
      <tr>
        <td>
          <a
            href="#"
          >
            a8prof
          </a>
        </td>
         
        <td>
          a8
        </td>
         
        <td>
          <i
            class="fa fa-bolt"
            title="consumption"
          />
           
          <span>
             current 
          </span>
           
          <!---->
           
          <!---->
           
          <span
            class="text-muted"
            title="period, average"
          >
            (8244µs, 4)
          </span>
           
          <!---->
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="#"
          >
            conso
          </a>
        </td>
         
        <td>
          m3
        </td>
         
        <td>
          <i
            class="fa fa-bolt"
            title="consumption"
          />
           
          <span>
             current 
          </span>
           
          <span>
             voltage 
          </span>
           
          <span>
             power 
          </span>
           
          <span
            class="text-muted"
            title="period, average"
          >
            (8244µs, 4)
          </span>
           
          <!---->
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="#"
          >
            radio
          </a>
        </td>
         
        <td>
          other
        </td>
         
        <td>
          <!---->
           
          <i
            class="fa fa-wifi"
            title="radio"
          />
           rssi
            
          <span
            class="text-muted"
            title="channel"
          >
            (11,13)
          </span>
        </td>
      </tr>
       
      <!---->
    </tbody>
  </table>
</div>
`;
